src/blockdata/block.rs

```diff
@@ -242,6 +243,42 @@ impl Block {
         let txs_weight: usize = self.txdata.iter().map(Transaction::get_weight).sum();
         base_weight + txs_weight
     }
+
+    /// Get the coinbase transaction, if one is present.
+    pub fn coinbase(&self) -> Option<&Transaction> {
+        self.txdata.iter().next()
+    }
+
+    /// Get the block height as encoded into the coinbase according to BIP34.
+    /// Returns [None] if not present.
+    pub fn bip34_block_height(&self) -> Option<u32> {
+        // Citing the spec:
+        // Add height as the first item in the coinbase transaction's scriptSig,
+        // and increase block version to 2. The format of the height is
+        // "serialized CScript" -- first byte is number of bytes in the number
+        // (will be 0x03 on main net for the next 150 or so years with 2^23-1
+        // blocks), following bytes are little-endian representation of the
+        // number (including a sign bit). Height is the height of the mined
+        // block in the block chain, where the genesis block is height zero (0).
+
+        if self.header.version < 2 {
+            return None;
+        }
+
+        self.coinbase().and_then(|cb| {
+            cb.input.iter().next().and_then(|input| {
+                input.script_sig.iter(true).next().and_then(|ins| match ins {
```

Comment:
I wasn't sure if this was needed. I don't think so, but the spec says "will be 0x03 on main net for the next 150 or so years" so I suppose we expect minimal pushes.

Reply:
Even though this will be ok for the next 80k years, maybe adding a case `script::Instruction::PushBytes(b) if b.len() > 4` which is marked as unreachable/panic would be more appropriate than just silently returning a wrong value?

Reply:
How so a wrong value? If it's over 4 we return `None` as in "there was no valid bip34 block height in the coinbase".

Reply:
I don't think the BIP actually states a maximum length for the encoded integer. For all we know it could become _really_ big (up to 2040 bit) :sweat_smile:

So it's probably best to just look [at Core](https://github.com/bitcoin/bitcoin/blob/34eb2362581d4d8f0bfd3baa12ba750afaf85c62/src/validation.cpp#L3553-L3560):

```cpp
    // Enforce rule that the coinbase starts with serialized block height
    if (nHeight >= consensusParams.BIP34Height)
    {
        CScript expect = CScript() << nHeight;
        if (block.vtx[0]->vin[0].scriptSig.size() < expect.size() ||
            !std::equal(expect.begin(), expect.end(), block.vtx[0]->vin[0].scriptSig.begin())) {
            return state.DoS(100, false, REJECT_INVALID, "bad-cb-height", false, "block height mismatch in coinbase");
        }
    }
```

Core allows minimally encoded, signed, 64bit values and so should we.

Reply:
Yes, I think we should return `u64` here

Reply:
fixing this